### PR TITLE
[TASK] Add generic sort operation

### PR DESCRIPTION
**Pull request by johannessteu**
This changes adds a generic sort operation that can be used to create any kind of sorting that ElasticSearch supports.

### CODE REVIEW

**Comment by johannessteu** [comment]
Waiting for #101 before this can get in

**Comment by aertmann** [comment]
Hey Johannes, you're submitting code that doesn't comply with PSR-2 which the package was changed to. Maybe you could maybe set your IDE to follow it or use the parser mentioned in https://discuss.neos.io/t/psr-2-change-for-flow-and-neos/493

**Comment by johannessteu** [comment]
Yeah, ide should already be on psr-2. I'll double check later why some changes are not made.

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryResult.php [hunk 1]
```diff
@@ -214,6 +214,23 @@ public function searchHitForNode(NodeInterface $node)
     }
 
     /**
+     * Returns the array with all sort values for a given node. The values are fetched from the raw content
+     * ElasticSearch returns within the hit data
+     *
+     * @param NodeInterface $node
+     * @return array
+     */
+    public function getSearchValuesForNode(NodeInterface $node)
```
Comment (daniellienert) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryResult.php]: The behavior of this method does not fit to its name ...
Do we really need this method at all?

**Comment by daniellienert** [comment]
@johannessteu I think the way to request specific meta data of a node by viewHelpers is not very intuitive. Wouldn't it be a better way to provide an array of nodes including their metaData like sort form  `ElasticSearchQueryResult::$result` and iterate over this array?

**Comment by johannessteu** [comment]
I think the main problem is the pagination. If you paginate over your ```ElasticSearchQueryResult``` you don't really have access to the result object. All iteration methods have to return an array with that node plus meta data. That would be a big breaking change. That's why there is the method getSearchValuesForNode, wich will be renamed, at all. There is no way to access the data. At least that is what @skurfuerst and i came up with. To get that data we thought about an ViewHelper since it is the only way to call a method and pass the current node to get meta data for that node.

I'll agree on the point that we could just provide one VH to return all meta data or with limit it with an argument to just one type or so. What do you think @daniellienert ?

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryResult.php [hunk 2]
```diff
@@ -214,6 +214,23 @@ public function searchHitForNode(NodeInterface $node)
...
+     * Returns the array with all sort values for a given node. The values are fetched from the raw content
+     * ElasticSearch returns within the hit data
+     *
+     * @param NodeInterface $node
+     * @return array
+     */
+    public function getSortValuesForNode(NodeInterface $node)
+    {
+        $hit = $this->searchHitForNode($node);
+        if (is_array($hit) && array_key_exists("sort", $hit)) {
```
Comment (aertmann) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryResult.php]: should be single quotes

**Comment by aertmann** [comment]
+1 regarding one VH instead of multiple specific ones

**Comment by daniellienert** [comment]
@johannessteu - ok I haven't thought about pagination. But isn't it possible to iterate over something like ElasticSearchQueryResult->RawResults and use RawResults.node to display that node?

If not, I am also in favor of having only one viewHelper.

**Comment by johannessteu** [comment]
I don't think so. It would be still at least a breaking change.
I'm not sure if it is possible to use the result as it is and set an iterator and use that interator.index to get data of an ElasticSearchQueryResult->RawResults. Still feels kind of nasty. I guess a single VH to render meta data for a node is way more flexible and could be used for some other use cases as well. Not just pagination. If you paginate over your result and below of above your list you like to show some meta data the VH could also be used. There might be a lot of use cases.

**Comment by daniellienert** [comment]
Ok, than lets do it with the viewHelper.

**Comment by johannessteu** [comment]
Ok, i'll try to get it in asap

**Comment by johannessteu** [comment]
@daniellienert @aertmann any ideas what exactly that one VH should return? The problem ist that all sort data will be inside the hit array, aggregations, suggestions etc. will be outside the hit array. What about to just keep the ```GetHitArrayForNodeViewHelper``` as sort data can be fetched over that?

**Comment by aertmann** [comment]
@johannessteu: I think that's what we both meant, remove ``GetSortValuesForNodeViewHelper`` and just use ``GetHitArrayForNodeViewHelper`` to get the sort data.

**Comment by daniellienert** [comment]
Thats correct - this generic `GetHitDataForNodeViewHelper` could i my opinion have a parameter "path" to address and output a specific value.

**Comment by aertmann** [comment]
would just return the array, then you can access ``{hit.sort}``

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/ViewHelpers/GetSortValuesForNodeViewHelper.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    /**
+     * @param ElasticSearchQueryResult $queryResultObject
+     * @param NodeInterface $node
+     * @return array
+     */
+    public function render(ElasticSearchQueryResult $queryResultObject, NodeInterface $node)
+    {
+        $sort = $queryResultObject->getSortValuesForNode($node);
+
+        if (count($sort) == 1) {
```
Comment (aertmann) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/ViewHelpers/GetSortValuesForNodeViewHelper.php]: ===

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/ViewHelpers/GetSortValuesForNodeViewHelper.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+use TYPO3\Fluid\Core\ViewHelper\AbstractViewHelper;
+
+/**
+ * View helper to get the raw "hits" array of an ElasticSearchQueryResult for a
+ * specific node.
+ *
+ * = Examples =
+ *
+ * <code title="Basic usage">
+ * {esCrAdapter:getSortValuesForNode(queryResultObject: {result}, node: {node})}
```
Comment (aertmann) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/ViewHelpers/GetSortValuesForNodeViewHelper.php]: syntax is incorrect, should be ``queryResultObject: result, node: node)``

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/ViewHelpers/GetHitArrayForNodeViewHelper.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+use TYPO3\Fluid\Core\ViewHelper\AbstractViewHelper;
+
+/**
+ * View helper to get the raw "hits" array of an ElasticSearchQueryResult for a
+ * specific node.
+ *
+ * = Examples =
+ *
+ * <code title="Basic usage">
+ * {esCrAdapter:geHitArrayForNode(queryResultObject: {result}, node: {node})}
```
Comment (aertmann) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/ViewHelpers/GetHitArrayForNodeViewHelper.php]: syntax is incorrect, should be ``queryResultObject: result, node: node)``

**Comment by johannessteu** [comment]
So here is the rework. I added the path thingy because you won't be really able to use {hit.sort}. You would have to use AliasView-VH or something. This might be more easy to use

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/ViewHelpers/GetHitArrayForNodeViewHelper.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+     * @param ElasticSearchQueryResult $queryResultObject
+     * @param NodeInterface $node
+     * @param string $path
+     * @return array
+     */
+    public function render(ElasticSearchQueryResult $queryResultObject, NodeInterface $node, $path = '')
+    {
+        $hitArray = $queryResultObject->searchHitForNode($node);
+
+        if (!empty($path) && array_key_exists($path, $hitArray)) {
```
Comment (daniellienert) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/ViewHelpers/GetHitArrayForNodeViewHelper.php]: I would suggest to use \TYPO3\Flow\Utility\Arrays::getValueByPath() to retrieve the array value.

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryBuilder.php [hunk 7]
```diff
@@ -171,15 +171,12 @@ public function nodeType($nodeType)
      */
     public function sortDesc($propertyName)
     {
-        if (!isset($this->request['sort'])) {
-            $this->request['sort'] = array();
-        }
-
-        // http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/search-request-sort.html
-        $this->request['sort'][] = array(
+        $configuration = array(
```
Comment (daniellienert) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryBuilder.php]: Use PHP 5.5 array structure?

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryBuilder.php [hunk 8]
```diff
@@ -193,19 +190,33 @@ public function sortDesc($propertyName)
      */
     public function sortAsc($propertyName)
     {
+        $configuration = array(
```
Comment (daniellienert) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryBuilder.php]: Use PHP 5.5 array structure?

File: Tests/Functional/Eel/ElasticSearchQueryTest.php [hunk 9]
```diff
@@ -158,9 +158,21 @@ public function nodesWillBeSortedDesc()
      */
     public function nodesWillBeSortedAsc()
     {
-        $ascendingResult = $this->queryBuilder->query($this->context->getRootNode())->sortAsc("title")->execute();
+        $ascendingResult = $this->queryBuilder->query($this->context->getRootNode())->sortAsc('title')->execute();
         $node = $ascendingResult->getFirst();
-        $this->assertEquals("chicken", $node->getProperty("title"), "Asserting a asc sort order by property title");
+        $this->assertEquals('chicken', $node->getProperty('title'), "Asserting a asc sort order by property title");
```
Comment (daniellienert) [inline Tests/Functional/Eel/ElasticSearchQueryTest.php]: There are still some doublequotes

File: Tests/Functional/Eel/ElasticSearchQueryTest.php [hunk 10]
```diff
@@ -148,9 +148,9 @@ public function fieldBasedAggregations()
      */
     public function nodesWillBeSortedDesc()
     {
-        $descendingResult = $this->queryBuilder->query($this->context->getRootNode())->sortDesc("title")->execute();
+        $descendingResult = $this->queryBuilder->query($this->context->getRootNode())->sortDesc('title')->execute();
         $node = $descendingResult->getFirst();
-        $this->assertEquals("egg", $node->getProperty("title"), "Asserting a desc sort order by property title");
+        $this->assertEquals('egg', $node->getProperty('title'), "Asserting a desc sort order by property title");
```
Comment (daniellienert) [inline Tests/Functional/Eel/ElasticSearchQueryTest.php]: There are still some doublequotes

File: Tests/Unit/ViewHelpers/GetHitArrayForNodeViewHelperTest.php [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+ *                                                                                                  */
+
+use Flowpack\ElasticSearch\ContentRepositoryAdaptor\Eel\ElasticSearchQueryResult;
+use Flowpack\ElasticSearch\ContentRepositoryAdaptor\ViewHelpers\GetHitArrayForNodeViewHelper;
+use TYPO3\TYPO3CR\Domain\Model\NodeInterface;
+
+/**
+ * Testcase for ElasticSearchQueryBuilder
+ */
+class GetHitArrayForNodeViewHelperTest extends \TYPO3\Flow\Tests\UnitTestCase
```
Comment (daniellienert) [inline Tests/Unit/ViewHelpers/GetHitArrayForNodeViewHelperTest.php]: Also use the PHP 5.5 array structure in the test?

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryBuilder.php [hunk 12]
```diff
@@ -171,14 +171,11 @@ public function nodeType($nodeType)
...
     public function sortDesc($propertyName)
     {
-        if (!isset($this->request['sort'])) {
-            $this->request['sort'] = array();
-        }
-
-        // http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/search-request-sort.html
-        $this->request['sort'][] = array(
+        $configuration = [
             $propertyName => array('order' => 'desc')
```
Comment (daniellienert) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryBuilder.php]: Mixed style makes it worse :)

Comment (johannessteu) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Eel/ElasticSearchQueryBuilder.php]: Well, we should go through all of the code at all in another PR to refactor all arrays or not?

**Pull request merged**